bundles/org.smarthomej.binding.tuya/src/main/java/org/smarthomej/binding/tuya/internal/util/IrUtils.java

```diff
@@ -0,0 +1,353 @@
+/**
+ * Copyright (c) 2021-2023 Contributors to the SmartHome/J project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.smarthomej.binding.tuya.internal.util;
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link IrUtils} is a support class for decode/encode infra-red codes
+ * <p>
+ * Based on https://github.com/jasonacox/tinytuya/blob/master/tinytuya/Contrib/IRRemoteControlDevice.py
+ *
+ * @author Dmitry P. (d51x) - Initial contribution
+ */
+public class IrUtils {
+    private static final Logger logger = LoggerFactory.getLogger(IrUtils.class);
+
+    private IrUtils() {
+    }
+
+    /**
+     * Convert Base64 code format from Tuya to nec-format.
+     *
+     * @param base64Code the base64 code format from Tuya
+     * @return the nec-format code
+     */
+    public static String base64ToNec(String base64Code) {
+        ArrayList<Integer> pulses = base64ToPulse(base64Code);
+        return pulsesToNec(pulses).get(0);
+    }
+
+    /**
+     * Convert Base64 code format from Tuya to samsung-format.
+     *
+     * @param base64Code the base64 code format from Tuya
+     * @return the samsung-format code
+     */
+    public static String base64ToSamsung(String base64Code) {
+        ArrayList<Integer> pulses = base64ToPulse(base64Code);
+        return pulsesToSamsung(pulses).get(0);
+    }
+
+    private static ArrayList<Integer> base64ToPulse(String base64Code) {
+        ArrayList<Integer> pulses = new ArrayList<>();
+        String key = (base64Code.length() % 4 == 1 && base64Code.startsWith("1")) ? base64Code.substring(1)
+                : base64Code;
+        byte[] raw_bytes = Base64.getDecoder().decode(key.getBytes(StandardCharsets.UTF_8));
+
+        int i = 0;
+        while (i < raw_bytes.length) {
+            int word = ((raw_bytes[i] & 0xFF) + (raw_bytes[i + 1] & 0xFF) * 256) & 0xFFFF;
+            pulses.add(word);
+            i += 2;
+        }
+        return pulses;
+    }
+
+    private static ArrayList<Long> pulsesToWidthEncoded(ArrayList<Integer> pulses, Integer start_mark,
```

Comment:
```suggestion
    private static List<Long> pulsesToWidthEncoded(List<Integer> pulses, Integer startMark,
```

use interfaces instead of implementation where possible and camel-case for variable and field names